client/fcidrClient.sp

```diff
@@ -0,0 +1,101 @@
+#include "SCR-Version"
+
+#include <sourcemod>
+#include <socket>
+#include <bytebuff>
+
+#define PROTOCOL_VERSION 1
+
+
+/**
+ * Base message structure
+ * 
+ */
+methodmap Header < ByteBuffer
+{
+	public Header()
+	{
+		return view_as<Header>(CreateByteBuffer());
+	}
+
+	public void baseHeader()
+	{
+		this.writeByte(PROTOCOL_VERSION);
+	}
+
+	public void Dispatch()
+	{
+		char sDump[MAX_BUFFER_LENGTH];
+
+		int iLen = this.Dump(sDump, MAX_BUFFER_LENGTH);
+
+		this.Close();
+
+		if (!SocketIsConnected(g_hSocket))
+			return;
+
+		// Len required
+		// If len is not included, it will stop at the first \0 terminator
+		SocketSend(g_hSocket, sDump, iLen);
+	}
+}
+
+methodmap Request < Header
+{
+	public Request(int ip_address, StringMap() keyvals)
+	{
+		BaseMessage mreq = BaseMessage();
+		mreq.baseHeader();
+		mreq.writeInt(ip_address);
+		mreq.writeInt(keyvals.Size);
+		for(int i = 0; i < keyvals.Snapshot().Length; i++){
+			keyvals.Snapshot().getKey(i, keyBuf);

```

Comment:
The snapshot should be assigned to a variable outside of the loop, continuous calling to `Snapshot()` will leak memory and does not guarantee the same order.

The single snapshot should also be closed when done.